crates/bevy_render/src/renderer/mod.rs

```diff
@@ -86,18 +86,120 @@ pub async fn initialize_renderer(
     #[cfg(not(feature = "wgpu_trace"))]
     let trace_path = None;
 
+    // Maybe get features and limits based on what is supported by the adapter/backend
+    let mut features = wgpu::Features::empty();
+    let mut limits = options.limits.clone();
     if matches!(options.priority, WgpuOptionsPriority::Functionality) {
-        let mut features =
-            adapter.features() | wgpu::Features::TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES;
+        features = adapter.features() | wgpu::Features::TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES;
         if adapter_info.device_type == wgpu::DeviceType::DiscreteGpu {
             // `MAPPABLE_PRIMARY_BUFFERS` can have a significant, negative performance impact for
             // discrete GPUs due to having to transfer data across the PCI-E bus and so it
             // should not be automatically enabled in this case. It is however beneficial for
             // integrated GPUs.
             features -= wgpu::Features::MAPPABLE_PRIMARY_BUFFERS;
         }
-        options.features = features;
-        options.limits = adapter.limits();
+        limits = adapter.limits();
+    }
+
+    // Enforce the disabled features
+    if let Some(disabled_features) = options.disabled_features {
+        features -= disabled_features;
+    }
+    // NOTE: |= is used here to ensure that any explicitly-enabled features are respected.
+    options.features |= features;
```

Comment:
I'm thinking that we should make WgpuOptions a "read only" configuration item. Users shouldn't be reading these settings to get "actual" values, they should be querying the device. And writing on top like this makes it hard to talk about what these values mean. Buuuut given that these changes are going into a patch release, we should probably hold off on that.

Reply:
I was thinking from the beginning that it’s a bit sketchy to reuse WgpuOptions to present the configured state after using it but it also isn’t used after init otherwise.

As regards this specific thing… we could have a private field with the final result or a separate resource so that WgpuOptions is used for configuration and WgpuCapabilities or some good name is what you can actually use. Both practically read only. But how do you even mark something as read-only? I didn’t know you could?

Reply:
In terms of "read only config", that ties into the conversation here: #2988
At the very least, I think it makes sense to "generally" treat plugin settings as "immutable after insertion", even though we don't have that functionality yet.

Reply:
plz also consider making plugin settings freeze when read https://discord.com/channels/691052431525675048/692572690833473578/941182628499947520 

This means you can:

* reconfigure settings before it is read by plugin
* disallow changing settings after it is read, so avoiding user confusion
* plugin can change to "reactive" settings without changing it's API